contracts/storage/ServiceAgreementStorage.sol

```diff
@@ -0,0 +1,503 @@
+    {
+        uint256 timeNow = block.timestamp;
+        ServiceAgreement storage agreement = serviceAgreements[agreementId];
+
+        ParametersStorage parametersStorage = ParametersStorage(hub.getContractAddress("ParametersStorage"));
+
+        uint256 proofWindowOffset = agreement.epochLength * agreement.epochsNum * agreement.proofWindowOffsetPerc / 100;
+        uint256 proofWindowDuration = agreement.epochLength * agreement.epochsNum * parametersStorage.proofWindowDurationPerc() / 100;
+
+        return (
+            timeNow > (agreement.startTime + agreement.epochLength * epoch + proofWindowOffset) &&
+            timeNow < (agreement.startTime + agreement.epochLength * epoch + proofWindowOffset + proofWindowDuration)
+        );
+    }
+
+    function getChallenge(address assetContract, uint256 tokenId, bytes memory keyword, uint8 hashingAlgorithm)
+        public
+        returns (bytes32, uint256)
+    {
+        bytes32 agreementId = _generateAgreementId(assetContract, tokenId, keyword, hashingAlgorithm);
+
+        IdentityStorage identityStorage = IdentityStorage(hub.getContractAddress("IdentityStorage"));
+        uint96 identityId = identityStorage.identityIds(keccak256(abi.encodePacked(msg.sender)));
+
+        AbstractAsset generalAssetInterface = AbstractAsset(assetContract);
+        bytes32 assertionId = generalAssetInterface.getAssertionByIndex(
+            tokenId,
+            AbstractAsset(assetContract).getAssertionsLength(tokenId) - 1
+        );
+
+        AssertionRegistry assertionRegistry = AssertionRegistry(hub.getContractAddress("AssertionRegistry"));
+        uint256 assertionSize = assertionRegistry.getSize(assertionId);
+
+        // blockchash() function only works for last 256 blocks (25.6 min window in case of 6s block time)
+        // TODO: figure out how to achieve randomness
+        return (
+            assertionId,
+            uint256(
+                sha256(abi.encodePacked(serviceAgreements[agreementId].proofWindowOffsetPerc, identityId))
+            ) % assertionSize
+        );
+    }
+
+    function sendProof(
+        address assetContract,
+        uint256 tokenId,
+        bytes memory keyword,
+        uint8 hashingFunctionId,
+        uint16 epoch,
+        bytes32[] memory proof,
+        bytes32 chunkHash
+    )
+        public
+    {
+        bytes32 agreementId = _generateAgreementId(assetContract, tokenId, keyword, hashingFunctionId);
+        require(!isProofWindowOpen(agreementId, epoch), "Proof window is open");
+
+        uint96 identityId = IdentityStorage(hub.getContractAddress("IdentityStorage")).identityIds(
+            keccak256(abi.encodePacked(msg.sender))
+        );
+
+        require(
+            commitSubmissions[keccak256(abi.encodePacked(agreementId, epoch, identityId))].score != 0,
+            "You have been already rewarded in this epoch"
+        );
+
+        bytes32 nextCommitId = serviceAgreements[agreementId].epochSubmissionHeads[epoch];
```

Comment:
we should be able to live without this variable as well (direcly call in the loop on line 367?)